Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx

```diff
@@ -0,0 +1,326 @@
+    name: 'Value',
+    fieldName: 'value',
+    minWidth: 100,
+    maxWidth: undefined,
+    isResizable: true,
+  },
+];
+
+const watchTableLayout: DetailsListLayoutMode = DetailsListLayoutMode.justified;
+
+// this can be exported and used in other places
+const getValueFromBotTraceScope = (delimitedProperty: string, botTrace: Activity) => {
+  const propertySegments = delimitedProperty.split('.');
+  const value = propertySegments.reduce(
+    (accumulator: object | string | number | boolean | undefined, segment, index) => {
+      // first try to grab the specified property off the root of the bot trace's memory
+      if (index === 0) {
+        return botTrace?.value[segment];
+      }
+      // if we are not on the root, try accessing the next value of the desired property
+      if (typeof accumulator === 'object') {
+        return accumulator[segment];
+      } else {
+        return undefined;
+      }
+    },
+    undefined
+  );
+  return value;
+};
+
+export const WatchTabContent: React.FC<DebugPanelTabHeaderProps> = ({ isActive }) => {
+  const currentProjectId = useRecoilValue(rootBotProjectIdSelector);
+  const rawWebChatTraffic = useRecoilValue(webChatTrafficState(currentProjectId ?? ''));
+  const [watchedVars, setWatchedVars] = useState<Record<string, string>>({});
+  const [pickerErrorMessages, setPickerErrorMessages] = useState<Record<string, string>>({});
+  const [selectedVars, setSelectedVars] = useState<IObjectWithKey[]>();
+  const [memoryVariablesPayload, setMemoryVariablesPayload] = useState<WatchDataPayload>({
+    kind: 'property',
+    data: { properties: [] },
+  });
+
+  const watchedVarsSelection = useMemo(
+    () =>
+      new Selection({
+        onSelectionChanged: () => {
+          setSelectedVars(watchedVarsSelection.getSelection());
+        },
+        selectionMode: SelectionMode.multiple,
+      }),
+    []
+  );
+
+  // get memory scope variables for the bot
+  useEffect(() => {
+    if (currentProjectId) {
+      const abortController = new AbortController();
+      (async () => {
+        try {
+          const variables = await getMemoryVariables(currentProjectId, { signal: abortController.signal });
+          setMemoryVariablesPayload({ kind: 'property', data: { properties: variables } });
+        } catch (e) {
+          // error can be due to abort
+        }
+      })();
+    }
+  }, [currentProjectId]);
+
+  const mostRecentBotState = useMemo(() => {
+    const botStateTraffic = rawWebChatTraffic.filter(
+      (t) => t.trafficType === 'activity' && t.activity.type === 'trace' && t.activity.name === 'BotState'
+    ) as ConversationActivityTrafficItem[];
+    if (botStateTraffic.length) {
+      return botStateTraffic[botStateTraffic.length - 1];
+    }
+  }, [rawWebChatTraffic]);
+
+  const onSelectPath = useCallback(
+    (variableId: string, path: string) => {
+      const watchedVar = Object.values(watchedVars).find((varPath) => varPath === path);
+      if (watchedVar) {
```

Comment:
have we considered a UX in which we dont show the selected vars in the list OR we disable them?

Reply:
1. What would be a scenario in which we don't show the selected variables in the list?

2. There are no designs at the moment for disabled variables

Reply:
when a variable is selected and they re-select it, it shows an error, which I think is too much, I think it should be silent about it, or if we consider it an "error" we can easily prevent it by not showing the selected variables in the select list or show them as disabled with a selected UI treatment (an explicit label or color etc.)

Reply:
Ahh I see. It might be the wording / naming of the code that is confusing you here.

"onSelectPath" is actually the "Submit" action of the picker. So let's say you have a row in the table that is already watching `user.myValue`. If you then add another row in the table, type in `user.myValue` and hit "Enter" you will see this validation error underneath the duplicate row and the duplicate **will not** be watched.

I spoke to @mewa1024 about this and we felt that this was a better UX than silently failing or doing a no-op.

Reply:
Here's an example:

![image](https://user-images.githubusercontent.com/3452012/122836097-78525d80-d2a6-11eb-8a56-7cb49e1be9a0.png)

Reply:
@mewa1024 @tonyanziano thanks for explaining the use case, it wasn't far from what I imagined. 👍🏼 

Chrome/edge dev tools have no issues with devs adding the same variable multiple times, I thought we may wanna consider following that.

Reply:
Ahhh, interesting! I wonder what VS / VS Code do in this situation.

Reply:
VS Code lets you do this without any feedback at all:

![image](https://user-images.githubusercontent.com/3452012/122838757-97072300-d2ab-11eb-9599-f02d36567520.png)

Reply:
another point I wanted to mention is captured in this screenshot above, "not available", if user enters something random that doesn't exist or they think it does but it doesn't, "not available" is a great indicator vs. undefined (if that's the case)
@mewa1024

Reply:
agree with all of the above. some other context about the red error message: I think Tony and I landed on that instead of just hiding the variable from the list because the user could type in something freeform (since it's a combobox-ish input rather than a pure dropdown input). 

About the "not available" versus "undefined", "not available" sounds good to me (I'll update the mocks for future reference.)

Reply:
@mewa1024 just to clarify, if a variable is indeed available but has no value at watch time, it would be undefined, but if it's not available (doesn't exist in the list, like user thinks it's there but it's not) it shows unavailable.

Reply:
ah, thanks for the clarification. in the JSON for the bot, would an undefined thing look like this:

`"propertyname" : "undefined"`

if so, seems like it should be shown in monospaced font, like this:
![image](https://user-images.githubusercontent.com/39317542/122969796-ed26a580-d341-11eb-9fdf-555481dfd8ac.png)

rather than this:
![image](https://user-images.githubusercontent.com/39317542/122969757-dc762f80-d341-11eb-8d4c-f34be59f1c79.png)

Thoughts?

Reply:
I think our editor only allows you to set a value to `null` if it is not set, if it is actually `undefined`, then I believe Monaco just omits the value from the display:

So here's what it would look like:

![image](https://user-images.githubusercontent.com/3452012/123307238-1c225000-d4d7-11eb-9b1b-7e6aa648b5d0.png)


If we had a `prop4: undefined` it would not even show up.